File: 105-从前序与中序遍历序列构建二叉树.py
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if not preorder:
            return None
        
        root = TreeNode(preorder[0])

        boundary = inorder.index(preorder[0])

        root.left = self.buildTree(preorder[1:boundary + 1], inorder[:boundary])
        root.right = self.buildTree(preorder[boundary + 1:], inorder[boundary + 1:])

        return root

File: 105-从前序与中序遍历序列构建二叉树.cpp
class Solution {
public:
    TreeNode* build_tree(vector<int>& preorder, vector<int>& inorder, int a, int b, int c, int d){
        if(a == b)
            return NULL;
        
        TreeNode* root = new TreeNode(preorder[a]);

        int i = c - 1;
        while(inorder[++i] != preorder[a]);

        root->left = build_tree(preorder, inorder, a + 1, a + i - c + 1, c, i);
        root->right = build_tree(preorder, inorder, a + i - c + 1, b, i + 1, d);

        return root;
    }
    
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return build_tree(preorder, inorder, 0, preorder.size(), 0, inorder.size());
    }
};
